File: 08-安全/02-网络安全/WireGuard握手/go/wg_handshake.go
// WireGuard 握手校验流程（main）：按 Noise_IKpsk2 的消息顺序逐步推进两端状态，
// 断言每一阶段 ck/h 一致、字段长度符合内核结构体、AEAD 篡改被拒、重放窗口正确。
//
// 运行:  go run .
package main

import (
	"crypto/rand"
	"fmt"
	"os"
)

func main() {
	check("LEN_INITIATION == 148", lenInitiation == 148, fmt.Sprint(lenInitiation))
	check("LEN_RESPONSE == 92", lenResponse == 92, fmt.Sprint(lenResponse))
	check("LEN_DATA_HEADER == 16", lenDataHeader == 16, fmt.Sprint(lenDataHeader))
	check("handshake_name 长度 37", len(handshakeName) == 37, fmt.Sprint(len(handshakeName)))
	check("identifier_name 长度 34", len(identifierName) == 34, fmt.Sprint(len(identifierName)))

	// BLAKE2s 官方向量（RFC 7693 附录 A 的 "abc"）
	want := "508c5e8c327c14e2e1a72ba34eeb452f37458b209ed63a294d999b4c86675982"
	got := fmt.Sprintf("%x", blake2s([]byte("abc"), nil, 32))
	check("BLAKE2s-256(\"abc\") 官方向量", got == want, got[:16])

	// 两侧各自的静态/临时密钥
	aPriv, aPub := x25519Keypair([]byte("alice-seed-0123456789abcdef012345"))
	bPriv, bPub := x25519Keypair([]byte("bob-seed-0123456789abcdef01234567"))
	ePriv, ePub := x25519Keypair([]byte("eph-seed-0123456789abcdef01234567"))
	e2Priv, e2Pub := x25519Keypair([]byte("eph2seed-0123456789abcdef0123456"))
	psk := make([]byte, symLen)
	for i := range psk {
		psk[i] = 0x33
	}

	// 发起方视角的 pre-message 是响应方静态公钥；响应方视角是自己那份
	syA := newSymmetric(bPub)
	syB := newSymmetric(bPub)
	check("ck0 一致", string(syA.ck) == string(syB.ck), "")
	check("h0 一致", string(syA.h) == string(syB.h), "")

	// msg1: e → es → s → ss
	syA.mixEphemeral(ePub)
	syB.mixEphemeral(ePub)
	esA, esB := x25519Shared(ePriv, bPub), x25519Shared(bPriv, ePub)
	check("DH(es) 两侧一致", string(esA) == string(esB), "")
	syA.mixKey(esA)
	syB.mixKey(esB)
	encStatic, _ := syA.encrypt(aPub)
	check("encrypted_static 长度 48", len(encStatic) == 48, fmt.Sprint(len(encStatic)))
	decStatic, okStatic := syB.decrypt(encStatic)
	check("响应方解出 encrypted_static == 发起方静态公钥",
		okStatic && string(decStatic) == string(aPub), "")
	ssA, ssB := x25519Shared(aPriv, bPub), x25519Shared(bPriv, aPub)
	check("DH(ss) 两侧一致", string(ssA) == string(ssB), "")
	syA.mixKey(ssA)
	syB.mixKey(ssB)
	encTs, _ := syA.encrypt(tai64n(1700000000, 0))
	check("encrypted_timestamp 长度 28", len(encTs) == 28, fmt.Sprint(len(encTs)))
	decTs, okTs := syB.decrypt(encTs)
	check("响应方解出时间戳（12 字节 TAI64N）",
		okTs && len(decTs) == tsLen && binary.BigEndian.Uint64(decTs) == 0x400000000000000A+1700000000, "")
	check("msg1 后 h 一致", string(syA.h) == string(syB.h), "")
	check("msg1 后 ck 一致", string(syA.ck) == string(syB.ck), "")

	// msg2: e → ee → se → psk → {}
	syA.mixEphemeral(e2Pub)
	syB.mixEphemeral(e2Pub)
	syA.mixKey(x25519Shared(ePriv, e2Pub))
	syB.mixKey(x25519Shared(e2Priv, ePub))
	seA, seB := x25519Shared(aPriv, e2Pub), x25519Shared(e2Priv, aPub)
	check("DH(se) 两侧一致", string(seA) == string(seB), "")
	syA.mixKey(seA)
	syB.mixKey(seB)
	syA.mixKeyAndHash(psk)
	syB.mixKeyAndHash(psk)
	check("psk 混入后 h 一致", string(syA.h) == string(syB.h), "")
	encNothing, _ := syB.encrypt(nil)
	check("encrypted_nothing 长度 16", len(encNothing) == tagLen, fmt.Sprint(len(encNothing)))
	decNothing, okNothing := syA.decrypt(encNothing)
	check("响应消息空负载认证通过", okNothing && len(decNothing) == 0, "")

	// split(): HKDF(ck, 空输入, 2) → 双向传输密钥
	kAB, kBA := syA.split()
	kAB2, kBA2 := syB.split()
	check("Split 两侧一致", string(kAB) == string(kAB2) && string(kBA) == string(kBA2), "")
	check("双向密钥不同", string(kAB) != string(kBA), "")

	// 传输数据消息：type|key_idx|counter|AEAD(空 AD)
	payload := []byte("hello wireguard")
	padded := append(append([]byte{}, payload...), make([]byte, (16-len(payload)%16)%16)...)
	dm := aeadSeal(kAB, append(make([]byte, 4), 7, 0, 0, 0, 0, 0, 0, 0), nil, padded)
	check("数据消息 = 16 + 16 + 16", lenDataHeader+len(dm) == 48, fmt.Sprint(lenDataHeader+len(dm)))
	opened, ok := aeadOpen(kAB, append(make([]byte, 4), 7, 0, 0, 0, 0, 0, 0, 0), nil, dm)
	check("数据消息解封成功", ok && len(opened) == 16, fmt.Sprint(len(opened)))
	bad := append([]byte{}, dm...)
	bad[len(bad)-1] ^= 0x01
	_, ok = aeadOpen(kAB, append(make([]byte, 4), 7, 0, 0, 0, 0, 0, 0, 0), nil, bad)
	check("篡改 tag 必须被拒", !ok, "")
	_, ok = aeadOpen(kBA, append(make([]byte, 4), 7, 0, 0, 0, 0, 0, 0, 0), nil, dm)
	check("反向密钥解不开", !ok, "")

	// MAC1 / cookie
	initMsg := make([]byte, lenInitiation)
	initMsg[0] = 1
	copy(initMsg[8:], ePub)
	_, otherPub := x25519Keypair([]byte("other-seed-0123456789abcdef0123456"))
	macBefore := computeMac1(initMsg, bPub)
	check("MAC1 绑响应方静态公钥", string(macBefore) != string(computeMac1(initMsg, otherPub)), "")
	check("MAC1 长度 16", len(macBefore) == cookieLen, fmt.Sprint(len(macBefore)))
	initMsg[8] ^= 0x01
	check("改一字节 → MAC1 变化", string(macBefore) != string(computeMac1(initMsg, bPub)), "")
	initMsg[8] ^= 0x01

	secret := make([]byte, symLen)
	for i := range secret {
		secret[i] = 0x77
	}
	srcIP := []byte{10, 0, 0, 1}
	c1 := makeCookie(secret, srcIP, 51820)
	c2 := makeCookie(secret, srcIP, 51821)
	check("cookie 绑定源端口", string(c1) != string(c2), "")

	// cookie 用 XChaCha20Poly1305（24 字节随机 nonce）+ MAC1 作为 AD
	nonce := make([]byte, 24)
	if _, err := rand.Read(nonce); err != nil {
		panic(err)
	}
	mac1 := computeMac1(initMsg, bPub)
	ckKey := blake2s(append(append([]byte{}, cookieLabel...), bPub...), nil, symLen)
	sealedCookie := xchacha20poly1305Seal(ckKey, nonce, mac1, c1)
	dec, ok := xchacha20poly1305Open(ckKey, nonce, mac1, sealedCookie)
	check("cookie 往返成功", ok && string(dec) == string(c1), "")
	_, ok = xchacha20poly1305Open(ckKey, nonce, mac1, sealedCookie[:len(sealedCookie)-1])
	check("cookie 截断必须被拒", !ok, "")

	// 重放窗口
	rc := newReplayCounter()
	check("首包接受", rc.validate(0), "")
	check("重复包拒绝", !rc.validate(0), "")
	check("跳窗接受", rc.validate(counterWindow+5), "")
	check("跳窗后旧号拒绝", !rc.validate(1), "")
	rc2 := newReplayCounter()
	check("边界：落后恰好一个窗口仍接受",
		rc2.validate(counterWindow) && rc2.validate(0), "")
	rc3 := newReplayCounter()
	check("边界：落后超一个窗口拒绝",
		rc3.validate(counterWindow+1) && !rc3.validate(0), "")

	// 定时器常量（内核 messages.h enum limits）
	check("REKEY_AFTER_TIME(120) < REJECT_AFTER_TIME(180)",
		rekeyAfterTime < rejectAfterTime, fmt.Sprintf("%d<%d", rekeyAfterTime, rejectAfterTime))
	check("MAX_TIMER_HANDSHAKES == 18", 90/rekeyTimeout == 18, fmt.Sprint(90/rekeyTimeout))
	check("KEEPALIVE_TIMEOUT == 10", keepaliveTime == 10, fmt.Sprint(keepaliveTime))

	fmt.Printf("\n%s: %d failure(s)\n", map[bool]string{true: "FAILED", false: "all checks passed"}[failures > 0], failures)
	if failures > 0 {
		os.Exit(1)
	}
}

File: 08-安全/02-网络安全/WireGuard握手/c/wg_handshake.c
// WireGuard 握手演示主程序（校验脚本）：编译
//   gcc -O2 -Wall -Wextra -pedantic wg_handshake.c -lcrypto -o wg_demo
// 实现分散在 wg_crypto_impl.h（BLAKE2s/HMAC/KDF）与 wg_noise_impl.h（X25519、AEAD、哈希链），
// 两者以文本级 #include 进本文件，因此构建命令无需列出它们。

#include "wg_noise_impl.h"

int main(void) {
    uint8_t a_priv[PUB_LEN], a_pub[PUB_LEN], b_priv[PUB_LEN], b_pub[PUB_LEN];
    uint8_t eph_priv[PUB_LEN], eph_pub[PUB_LEN], es[PUB_LEN], ss[PUB_LEN], ee[PUB_LEN];
    uint8_t psk[SYM_LEN], mac1_key[SYM_LEN];
    Symmetric sy_a, sy_b;
    uint8_t label[8 + PUB_LEN];

    if (!x25519_keypair(a_priv, a_pub) || !x25519_keypair(b_priv, b_pub) ||
        !x25519_keypair(eph_priv, eph_pub)) {
        fprintf(stderr, "X25519 keygen failed\n");
        return 1;
    }
    memset(psk, 0x33, sizeof(psk));

    /* --- 协议常量与消息长度 --- */
    check("LEN_INITIATION == 148", LEN_INITIATION == 148, LEN_INITIATION, 148);
    check("LEN_RESPONSE == 92", LEN_RESPONSE == 92, LEN_RESPONSE, 92);
    check("LEN_DATA_HDR == 16", LEN_DATA_HDR == 16, LEN_DATA_HDR, 16);
    check("handshake_name 长度 37",
          sizeof(HANDSHAKE_NAME) - 1 == 37, (long)(sizeof(HANDSHAKE_NAME) - 1), 37);
    check("identifier_name 长度 34",
          sizeof(IDENTIFIER_NAME) - 1 == 34, (long)(sizeof(IDENTIFIER_NAME) - 1), 34);

    /* --- BLAKE2s 官方向量：BLAKE2s-256("abc") --- */
    {
        static const uint8_t want[HASH_LEN] = {
            0x50, 0x8C, 0x5E, 0x8C, 0x32, 0x7C, 0x14, 0xE2, 0xE1, 0xA7, 0x2B, 0xA3,
            0x4E, 0xEB, 0x45, 0x2F, 0x37, 0x45, 0x8B, 0x20, 0x9E, 0xD6, 0x3A, 0x29,
            0x4D, 0x99, 0x9B, 0x4C, 0x86, 0x67, 0x59, 0x82
        };
        uint8_t got[HASH_LEN];
        int i, diff = 0;
        blake2s(got, HASH_LEN, NULL, 0, (const uint8_t *)"abc", 3);
        for (i = 0; i < HASH_LEN; i++) diff |= got[i] ^ want[i];
        check("BLAKE2s-256(\"abc\") 官方向量", diff == 0, diff, 0);
    }

    /* --- 两侧各自推进 msg1：ck/h 必须逐步一致 --- */
    handshake_init(&sy_a, b_pub);   /* 发起方视角：pre-message 是响应方静态公钥 */
    handshake_init(&sy_b, b_pub);   /* 响应方视角：pre-message 是自己的静态公钥 */
    check("ck0 相等", memcmp(sy_a.ck, sy_b.ck, HASH_LEN) == 0, 0, 0);
    check("h0 相等", memcmp(sy_a.h, sy_b.h, HASH_LEN) == 0, 0, 0);

    /* e：发起方临时公钥（PSK 握手还要用同一公钥做一次 MixKey(ck)） */
    mix_hash(&sy_a, eph_pub, PUB_LEN);
    mix_key_only_ck(&sy_a, eph_pub, PUB_LEN);
    mix_hash(&sy_b, eph_pub, PUB_LEN);
    mix_key_only_ck(&sy_b, eph_pub, PUB_LEN);
    check("e token 后 h 相等", memcmp(sy_a.h, sy_b.h, HASH_LEN) == 0, 0, 0);

    /* es / ss：两侧用不同私钥算同一个 DH */
    if (!x25519(es, eph_priv, b_pub) || !x25519(ss, a_priv, b_pub)) {
        fprintf(stderr, "X25519 derive failed\n");
        return 1;
    }
    {
        uint8_t es_b[PUB_LEN], ss_b[PUB_LEN];
        x25519(es_b, b_priv, eph_pub);
        x25519(ss_b, b_priv, a_pub);
        check("DH(es) 两侧一致", memcmp(es, es_b, PUB_LEN) == 0, 0, 0);
        check("DH(ss) 两侧一致", memcmp(ss, ss_b, PUB_LEN) == 0, 0, 0);
    }
    mix_key(&sy_a, es, PUB_LEN);
    mix_key(&sy_b, es, PUB_LEN);
    /* s 字段被加密（48 字节 = 32 + 16 tag），这里只验证长度与混入 h */
    {
        uint8_t ct[PUB_LEN + TAG_LEN];
        aead_seal(ct, sy_a.k, sy_a.h, HASH_LEN, a_pub, PUB_LEN);
        check("encrypted_static 长度 48", sizeof(ct) == 48, (long)sizeof(ct), 48);
        mix_hash(&sy_a, ct, sizeof(ct));
        mix_hash(&sy_b, ct, sizeof(ct));
    }
    mix_key(&sy_a, ss, PUB_LEN);
    mix_key(&sy_b, ss, PUB_LEN);
    check("msg1 结束时 h 相等", memcmp(sy_a.h, sy_b.h, HASH_LEN) == 0, 0, 0);
    check("msg1 结束时 ck 相等", memcmp(sy_a.ck, sy_b.ck, HASH_LEN) == 0, 0, 0);

    /* --- 用 h 作为 AD 造一条 12 字节 TAI64N 时间戳（时间只影响明文，不影响链） */
    {
        uint8_t ts[TS_LEN], sealed_ts[TS_LEN + TAG_LEN];
        uint64_t now = 0x400000000000000AULL + 1700000000ULL;
        int i;
        for (i = 0; i < 8; i++) ts[i] = (uint8_t)(now >> (8 * (7 - i)));
        memset(ts + 8, 0, 4);
        aead_seal(sealed_ts, sy_a.k, sy_a.h, HASH_LEN, ts, TS_LEN);
        check("encrypted_timestamp 长度 28", sizeof(sealed_ts) == 28,
              (long)sizeof(sealed_ts), 28);
        mix_hash(&sy_a, sealed_ts, sizeof(sealed_ts));
        mix_hash(&sy_b, sealed_ts, sizeof(sealed_ts));
    }

    /* --- msg2：e → ee → se → psk → {} --- */
    {
        uint8_t eph2_priv[PUB_LEN], eph2_pub[PUB_LEN], ct[TAG_LEN];
        uint8_t newk_a[SYM_LEN], newk_b[SYM_LEN], ck_split_a[HASH_LEN], ck_split_b[HASH_LEN];
        x25519_keypair(eph2_priv, eph2_pub);
        mix_hash(&sy_b, eph2_pub, PUB_LEN);
        mix_key_only_ck(&sy_b, eph2_pub, PUB_LEN);
        mix_hash(&sy_a, eph2_pub, PUB_LEN);
        mix_key_only_ck(&sy_a, eph2_pub, PUB_LEN);
        if (!x25519(ee, eph2_priv, eph_pub)) return 1;
        mix_key(&sy_b, ee, PUB_LEN);
        mix_key(&sy_a, ee, PUB_LEN);
        {   /* se：响应方用自方临时私钥 + 发起方静态公钥 */
            uint8_t se_b[PUB_LEN], se_a[PUB_LEN];
            x25519(se_b, eph2_priv, a_pub);
            x25519(se_a, a_priv, eph2_pub);
            check("DH(se) 两侧一致", memcmp(se_b, se_a, PUB_LEN) == 0, 0, 0);
            mix_key(&sy_b, se_b, PUB_LEN);
            mix_key(&sy_a, se_a, PUB_LEN);
        }
        mix_key_and_hash(&sy_b, psk);
        mix_key_and_hash(&sy_a, psk);
        check("psk 混入后 h 相等", memcmp(sy_a.h, sy_b.h, HASH_LEN) == 0, 0, 0);
        aead_seal(ct, sy_b.k, sy_b.h, HASH_LEN, NULL, 0);
        check("encrypted_nothing 长度 16", sizeof(ct) == TAG_LEN, (long)sizeof(ct), 16);
        mix_hash(&sy_b, ct, TAG_LEN);
        mix_hash(&sy_a, ct, TAG_LEN);
        /* split(): HKDF(ck, 空输入, 2) → 双向传输密钥 */
        kdf(newk_a, newk_b, NULL, NULL, 0, sy_a.ck);
        memcpy(ck_split_a, sy_a.ck, HASH_LEN);
        memcpy(ck_split_b, sy_b.ck, HASH_LEN);
        check("Split 前 ck 相等", memcmp(ck_split_a, ck_split_b, HASH_LEN) == 0, 0, 0);
        check("两个方向密钥不同", memcmp(newk_a, newk_b, SYM_LEN) != 0, 0, 0);
        /* 传输消息：KEY_IDX + 计数器 + 密文 */
        {
            uint8_t dm[LEN_DATA_HDR + 32], pt[16];
            size_t i;
            memset(pt, 0xAB, sizeof(pt));
            dm[0] = 4; dm[1] = dm[2] = dm[3] = 0;
            for (i = 0; i < 8; i++) dm[8 + i] = (uint8_t)(i == 0 ? 7 : 0);
            aead_seal(dm + LEN_DATA_HDR, newk_a, NULL, 0, pt, sizeof(pt));
            check("数据消息总长 = 16+16+16", sizeof(dm) == 48, (long)sizeof(dm), 48);
        }
    }

    /* --- MAC1：密钥 = BLAKE2s("mac1----" ‖ 响应方静态公钥)，输出 16 字节 --- */
    memcpy(label, MAC1_LABEL, 8);
    memcpy(label + 8, b_pub, PUB_LEN);
    blake2s(mac1_key, SYM_LEN, NULL, 0, label, sizeof(label));
    {
        uint8_t init_msg[LEN_INITIATION];
        uint8_t other_key[SYM_LEN], label2[8 + PUB_LEN];
        uint8_t other_priv[PUB_LEN], other_pub[PUB_LEN];
        uint8_t mac_right[COOKIE_LEN], mac_wrong[COOKIE_LEN], mac_tampered[COOKIE_LEN];

        memset(init_msg, 0, sizeof(init_msg));
        init_msg[0] = 1;                        /* type = HANDSHAKE_INITIATION */
        memcpy(init_msg + 8, eph_pub, PUB_LEN); /* 这里只求确定字节串，不做真加密 */
        blake2s(mac_right, COOKIE_LEN, mac1_key, SYM_LEN,
                init_msg, LEN_INITIATION - MACS_LEN);

        x25519_keypair(other_priv, other_pub);  /* 换一个「响应方」静态公钥 */
        memcpy(label2, MAC1_LABEL, 8);
        memcpy(label2 + 8, other_pub, PUB_LEN);
        blake2s(other_key, SYM_LEN, NULL, 0, label2, sizeof(label2));
        blake2s(mac_wrong, COOKIE_LEN, other_key, SYM_LEN,
                init_msg, LEN_INITIATION - MACS_LEN);

        init_msg[8] ^= 0x01;                    /* 改一个字节 */
        blake2s(mac_tampered, COOKIE_LEN, mac1_key, SYM_LEN,
                init_msg, LEN_INITIATION - MACS_LEN);

        check("MAC1 长度 16", COOKIE_LEN == 16, COOKIE_LEN, 16);
        check("换响应方公钥 → MAC1 不同",
              memcmp(mac_right, mac_wrong, COOKIE_LEN) != 0, 0, 0);
        check("改消息一个字节 → MAC1 不同",
              memcmp(mac_right, mac_tampered, COOKIE_LEN) != 0, 0, 0);
    }
    /* cookie：keyed-BLAKE2s(secret, 源IP ‖ 源端口)，换端口即不同 */
    {
        uint8_t secret[SYM_LEN], c1[COOKIE_LEN], c2[COOKIE_LEN], buf[6];
        memset(secret, 0x77, sizeof(secret));
        memset(buf, 0, sizeof(buf));
        buf[3] = 1;
        buf[4] = 0xCA; buf[5] = 0x6C;               /* 端口 51820 */
        blake2s(c1, COOKIE_LEN, secret, SYM_LEN, buf, sizeof(buf));
        buf[5] = 0x6D;                              /* 端口 51821 */
        blake2s(c2, COOKIE_LEN, secret, SYM_LEN, buf, sizeof(buf));
        check("cookie 绑定源端口", memcmp(c1, c2, COOKIE_LEN) != 0, 0, 0);
    }
    /* cookie 加密密钥：BLAKE2s("cookie--" ‖ 响应方静态公钥)，与 MAC1 密钥同构但不同 */
    {
        uint8_t ck_key[SYM_LEN], label3[8 + PUB_LEN];
        memcpy(label3, COOKIE_LABEL, 8);
        memcpy(label3 + 8, b_pub, PUB_LEN);
        blake2s(ck_key, SYM_LEN, NULL, 0, label3, sizeof(label3));
        check("cookie 密钥 ≠ MAC1 密钥", memcmp(ck_key, mac1_key, SYM_LEN) != 0, 0, 0);
    }

    printf("\n%s: %d failure(s)\n", failures ? "FAILED" : "all checks passed", failures);
    return failures ? 1 : 0;
}
